### ISSUE TITLE
Timeouts and such

### ISSUE THREAD

**Issue by Martii** [Automattic/mongoose#7355]
*Before creating an issue please make sure you are using the latest version of mongoose*

We are. [npmjs.com with 5.4.0](https://www.npmjs.com/package/mongoose)

**Do you want to request a *feature* or report a *bug*?**

Potential bug... probably somewhere inbetween AWS/MongoLabs and MongoDB/*mongoose*

**What is the current behavior?**
OpenUserJS/OpenUserJS.org#1548 ... general instability with DB's that appeared out of nowhere.

**If the current behavior is a bug, please provide the steps to reproduce.**

See OpenUserJS/OpenUserJS.org#1548 for extremely detailed debug session over the last few days. Most relevant for *mongoose* is [here](https://github.com/OpenUserJS/OpenUserJS.org/issues/1548#issuecomment-448472671).

Summary... do a DB call and the site lags intermittently and mostly constantly which cascades to other dependencies... but a few breaks here and there. See https://openuserjs.org . Don't do a DB call and most of the time it follows through but not always. See https://openuserjs.org/about

**What is the expected behavior?**
Unsure ... Level3 backbone is having issues, so is our VPS, so is *async* dependency and so is *mongoose* via the driver. A few days back it was fine and we didn't even have any updates in the works. I can only assume that AWS/MongoLabs is having at least some issues. I don't have access to those accounts as our establishing owner does that part.

**Please mention your node.js, mongoose and MongoDB version.**

*node*@10.14.2
*mongodb*@3.1.10
MongoDB v3.2.20 pro(duction) to AWS/MongoLabs *(please remember I can't change this since no privi's)*
MongoDB v3.6.6 dev(elopment) to AWS/MongoLabs *(please remember I can't change this since no privi's)*

Recent changes, specific to *mongoose*, to try to improve connection in our project:
* https://github.com/OpenUserJS/OpenUserJS.org/commit/544c9fbaf7d5203b7f39d9c44c017c0bf6302b8a *(Failed... see revert below)*
* https://github.com/OpenUserJS/OpenUserJS.org/commit/0924e000d110ac49b6d30dae3e7a521b4fdb0d51
* https://github.com/OpenUserJS/OpenUserJS.org/commit/a32e8bc29d29817b7a6aa3b296719f0d151d74ad
* https://github.com/OpenUserJS/OpenUserJS.org/commit/ba9852661a11caed0cbd0693f5e808d3f837a7ed *(Revert to mongoose default)*

There's a ton of information at OpenUserJS/OpenUserJS.org#1548 regarding other dependency and VPS... please read there.

---

* Current `dbOptions` [~~here~~](https://github.com/OpenUserJS/OpenUserJS.org/blob/a32e8bc29d29817b7a6aa3b296719f0d151d74ad/app.js#L68-L83) [here](https://github.com/OpenUserJS/OpenUserJS.org/blob/ba9852661a11caed0cbd0693f5e808d3f837a7ed/app.js#L68-L83)
* Connection [here](https://github.com/OpenUserJS/OpenUserJS.org/blob/a32e8bc29d29817b7a6aa3b296719f0d151d74ad/app.js#L96)

Any insight would be greatly appreciated.

---
NOTE: I just took the site offline *(parked)* since we're getting no content with success responses from AWS. This is very bad for .user.js as it could overwrite their existing scripts. :\ HELP! please. :)

---

NOTE: Now we're in Read Only mode... so can see what is going on.

**Comment by vkarpov15** [comment]
Have you tried increasing your socketTimeoutMS? https://docs.mongodb.com/manual/reference/connection-string/#connection-options . It looks like you're experiencing network latency between your app servers and the mongodb server, so sockets might be timing out.

**Comment by Martii** [comment]
> Have you tried increasing your socketTimeoutMS?

Sure haven't.

What is the default value. The https://docs.mongodb.com/manual/reference/connection-string/#connection-options takes me to a generic page. Searching on their site for it currently yields no results. See https://docs.mongodb.com/ecosystem/drivers/?searchProperty=current&query=socketTimeoutMS

**Comment by Martii** [comment]
One addition... MongoDB docs say it never times out e.g.

> "The default is never to timeout".

So I don't think increasing it would help I think. That would actually be a decrease from infinity. :\

**Comment by Martii** [comment]
> It looks like you're experiencing network latency between your app servers and the mongodb server, so sockets might be timing out.

With `socketTimeoutMS` at https://mongoosejs.com/docs/connections.html it says *mongoose* sets it to 30000.

I'll try bumping it up 2x or 3x but not sure if increasing will help it. 

It's a weird issue. I've tempered it with *node*@11.x a bit but it still exhibits the symptoms. 

Assuming it's the holiday traffic this issue might just disappear by itself but the integrity would be still diminished with perception.

**Comment by Martii** [comment]
2x failed... test, clicking around site, then pause, then click... stalled e.g. web spinner continuing and no content sent. Nice when tests give me results, even if not what I was hoping for,  quickly. ;)

**Comment by vkarpov15** [comment]
This really just sounds like network latency. Not sure what I can do to help without access to your production setup :\ Another piece of advice would be to look out for slow queries and bad index misses - what does your query targeting panel in MongoDB cloud manager look like?

**Comment by Martii** [comment]
> This really just sounds like network latency.

From your keyboard thoughts to a higher powers eyes! :smile_cat:  mLabs shows definite lag that corresponds to my manual testing and one automated using *clinic* ... see graph at https://github.com/OpenUserJS/OpenUserJS.org/issues/1548#issuecomment-451192512 from mLabs.

> look out for slow queries

mLabs says we have a few, with some recommendations, but with as little UI traffic as we really get this shouldn't be a huge issue just yet. I'm still analyzing what mLabs gives to me via the UI and will improve those over time.

I just upgraded mLabs to MongoDB 3.6.9 *(which is production)*... still working on getting dev(elopment) MongoDB up to that version...installing this "old" of a MongoDB version in a newer distro is proving to be challenging especially since the db folder moved. It seems like mLabs will only let me download the db and do conversion locally and upload it back for development. Still investigating this procedure. I'd rather have this all sync'd on the same version between pro, dev, and local... so things are a little more consistent across the board with testing. Eventually we may have to move the DB to our VPS to speed this up... although that will just cover up the *async*/*node* dependency issue found.

> and bad index misses

The UI doesn't show any of those in a quick scan but has a lot of recommendations for building more indexes. Overall the general graphs mLabs provide seem nominal compared to some usage tests in local MongoDB graph set.

> what does your query targeting panel in MongoDB cloud manager look like?

If that's via `mongo` shell then I haven't gotten that far yet with our production DB. We do have some "Page Faults (per second)" over the last two weeks. That's something I'm not familiar with just yet.

> Another piece of advice...

I appreciate anything. I'm ~2 hours into mLabs. LOL but have more under my belt for MongoDB locally. :smiley:

**Comment by Martii** [comment]
Should I be worried about the "Page Faults (per second)"? Still unsure what this means in relation to MongoDB and mLabs.

**Comment by vkarpov15** [comment]
AFAIK elevated page faults means that your mongodb server is thrashing: https://en.wikipedia.org/wiki/Thrashing_(computer_science) . That isn't very good, and likely means you can solve this problem by increasing the amount of RAM on the machine your MongoDB server is running on.

**Comment by Martii** [comment]
Okee dokee.

> likely means you can solve this problem by increasing the amount of RAM on the machine your MongoDB server is running on.

Well with mLabs this option doesn't seem possible. The replica set is done with a VM on their end that we don't appear to have any control of it.

> elevated page faults

See also: https://blog.mlab.com/2015/08/telemetry-series-page-faults/

So in short... not enough RAM, too many indexes, and/or too little indexes. e.g. I know what they'll end up saying... optimize your indexes which of course is only a small part of the total issue. ;) 

We will work on that however it's still possible that we won't be using mLabs anymore since the network lag/latency is still present albeit on a declining slope over the last few days since I migrated to 3.6.9 and perhaps lower user interaction. Still investigating what I can do on the systems end before I dig into optimizing indexes. mLabs allows index setting directly in their UI and I read that [*mongoose* supports secondary indexes](https://mongoosejs.com/docs/guide.html#indexes). I'd like to keep most of the indexes in our source as much as possible so we can tweak these on the fly instead of just a few of us who have access on mLabs *(or perhaps eventually a local production MongoDB)*.

Looking back on the stats over the last year we didn't seem to have any issues when running on the replica DB *(the repolicated DB backup in the replica set)* as primary between March and September 2018... which is a little odd but just noticed that. I didn't toggle the "PRIMARY" *(which I assume is master)* as I just got access to mLabs several days ago.

Anyhow... I really appreciate your feedback. BIG THANK YOU! :smiley:  I'll figure out how to issue *async* dependency and perhaps add some more error traps in the following days since it fails with this intermittent network lag. e.g. when I de-*async*'d a controller route in OpenUserJS/OpenUserJS.org#1562 that page immediately sped up. It does have a *mongoose* DB call but when used with the *async* dependency using callbacks, or `async`/`await` ES6+ syntax alternatively, in combination with the lag/latency the page fails miserably when lag/latency encountered.
